src/ImageSharp/Formats/Png/Filters/PaethFilter.cs

```diff
@@ -127,5 +154,36 @@ private static byte PaethPredictor(byte left, byte above, byte upperLeft)
 
             return upperLeft;
         }
+
+        private static Vector<byte> PaethPredictor(Vector<byte> left, Vector<byte> above, Vector<byte> upperLeft)
+        {
+            Vector.Widen(left, out Vector<ushort> a1, out Vector<ushort> a2);
+            Vector.Widen(above, out Vector<ushort> b1, out Vector<ushort> b2);
+            Vector.Widen(upperLeft, out Vector<ushort> c1, out Vector<ushort> c2);
+
+            Vector<short> p1 = PaethPredictor(Vector.AsVectorInt16(a1), Vector.AsVectorInt16(b1), Vector.AsVectorInt16(c1));
+            Vector<short> p2 = PaethPredictor(Vector.AsVectorInt16(a2), Vector.AsVectorInt16(b2), Vector.AsVectorInt16(c2));
+            return Vector.AsVectorByte(Vector.Narrow(p1, p2));
+        }
```

Comment:
I worked out a way to implement Paeth without widening/narrowing as well.  Assuming `left`, `above`, and `upleft` are `Vector256<byte>` and `zero` is `Vector256<byte>.Zero`:

```C#
// Here, we refactor pa = abs(p - left) = abs(left + above - upleft - left)
// to pa = abs(above - upleft).  Same deal for pb.
// Using saturated subtraction, if the result is negative, the output is zero.
// If we subtract in both directions and `or` the results, only one can be
// non-zero, so we end up with the absolute value.
var sac = Avx2.SubtractSaturate(above, upleft);
var sbc = Avx2.SubtractSaturate(left, upleft);
var pa = Avx2.Or(Avx2.SubtractSaturate(upleft, above), sac);
var pb = Avx2.Or(Avx2.SubtractSaturate(upleft, left), sbc);

// pc = abs(left + above - upleft - upleft), or abs(left - upleft + above - upleft).
// We've already calculated left - upleft and above - upleft in `sac` and `sbc`.
// If they are both negative or both positive, the absolute value of their
// sum can't possibly be less than `pa` or `pb`, so we'll never use the value.
// We make a mask that sets the value to 255 if they either both got
// saturated to zero or both didn't.  Then we calculate the absolute value
// of their difference using saturated subtract and `or`, same as before,
// keeping the value only where the mask isn't set.
var pm = Avx2.CompareEqual(Avx2.CompareEqual(sac, zero), Avx2.CompareEqual(sbc, zero));
var pc = Avx2.Or(pm, Avx2.Or(Avx2.SubtractSaturate(pb, pa), Avx2.SubtractSaturate(pa, pb)));

// Finally, blend the values together.  We start with `upleft` and overwrite on
// tied values so that the `left`, `above`, `upleft` precedence is preserved.
var minbc = Avx2.Min(pc, pb);
var resbc = Avx2.BlendVariable(upleft, above, Avx2.CompareEqual(minbc, pb));
return Avx2.BlendVariable(resbc, left, Avx2.CompareEqual(Avx2.Min(minbc, pa), pa));
```

Haven't benchmarked it, but it's about half the work, so...

Reply:
Looks very promising. I obviously did everything in the easiest way possible (and because I'm not great at the specialized instructions). I'll write a benchmark for it and add it before the Vector path.

Reply:
🤯 
![image](https://user-images.githubusercontent.com/23627133/118611084-c2828380-b7bc-11eb-9f82-9ff3ccdcf25b.png)

Reply:
Can I use this code in my MIT imaging library too? 😅 
https://github.com/TechnologicalPizza/StbSharp.Image

Reply:
Of course you can!  It was a fun challenge 😄

Reply:
Would love to see some proper benchmarks in that repo. I reckon our jpeg decoder will be much closer now! 😛